File: benchmarks/proto_bench.cc
// ProtoBench.java's cpp version
// Copyright enjolras1205@gmail.com
// Feel free to use, reuse and abuse the code in this file.

#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#include <iostream>
#include <fstream>
#include <sstream>
#include <functional>

#include <google/protobuf/io/coded_stream.h>
#include <google/protobuf/io/zero_copy_stream_impl.h>

#include <google_speed.pb.h>
#include <google_size.pb.h>

using namespace google::protobuf::io;
using namespace std;

#define RETURN_IF_FAILED(x) do { int ret = (x); if (ret != 0) return ret; } while(0)
#define GO_ERROR_IF_FAILED(x) do { int ret = (x); if (ret != 0) goto error; } while(0)

const long MIN_SAMPLE_TIME_US = 2 * 1000000;
const long TARGET_TIME_US = 30 * 1000000;

typedef function<int(void)> Func;

// calc usec gap of start to now
int time_action(int iterations, const Func &func, long &gap) {
	struct timeval start;
	struct timeval end;
	gettimeofday(&start, 0);
	for (int i = 0; i < iterations; ++i) {
		if (!func()) {
			cerr << "serialize/deserialize error" << endl;
			return -1;
		}
	}
	gettimeofday(&end, 0);
	gap = (end.tv_sec - start.tv_sec) * 1000000 + end.tv_usec - start.tv_usec;
	return 0;
}

int benchmark(const string &test_type, size_t data_size, const Func &func) {
	// Run it progressively more times until we've got a reasonable sample
	int iterations = 1;
	long elapsed = 0;
	RETURN_IF_FAILED(time_action(iterations, func, elapsed));
	while (elapsed < MIN_SAMPLE_TIME_US) {
		iterations *= 2;
		RETURN_IF_FAILED(time_action(iterations, func, elapsed));
	}
	// Upscale the sample to the target time. Do this in floating point arithmetic
	// to avoid overflow issues.
	iterations = (int) ((TARGET_TIME_US / (double) elapsed) * iterations);
	RETURN_IF_FAILED(time_action(iterations, func, elapsed));

	cout<< test_type << iterations << " iterations in "
			<< elapsed / 1000000.0 << "s; "
			<< (static_cast<double>(iterations) * data_size) / (elapsed / 1000000.0 * 1024 * 1024)
			<< "MB/s" <<endl;

	return 0;
}

template<typename MSG>
int run_test(const string &file_path, const string &msg_type) {
	ifstream file_stream(file_path);

	if (!file_stream) {
		cerr << "can't read from file " << file_path << endl;
		return -1;
	}

	int fd = open("/dev/null", O_WRONLY);
	if (fd < 0) {
		cerr << "can't open dev null for write" << endl;
		return -1;
	}

	file_stream.seekg(0, std::ios::end);
	size_t data_size = file_stream.tellg();

	// str
	string str;
	str.reserve(data_size);
	file_stream.seekg(0, std::ios::beg);
	str.assign((std::istreambuf_iterator<char>(file_stream)), std::istreambuf_iterator<char>());

	// byte array
	char *arr = new char[data_size];

	// memory out stream
	ostringstream o_str_stream;
	OstreamOutputStream o_mem_stream(&o_str_stream);

	// dev null out stream
	FileOutputStream null_stream(fd);
	// reuse dev null out stream
	CodedOutputStream reuse_null_stream(&null_stream);

	// memory in stream
	istringstream i_str_stream;
	IstreamInputStream i_mem_stream(&i_str_stream);

	MSG msg;

	GO_ERROR_IF_FAILED(!msg.ParseFromString(str));

	GO_ERROR_IF_FAILED(benchmark(msg_type + "Serialize to byte string ", data_size,
		[&msg, &str]() -> int { return msg.SerializeToString(&str); }));
	GO_ERROR_IF_FAILED(benchmark(msg_type + "Serialize to byte array ", data_size,
		[&msg, arr, data_size]() -> int { return msg.SerializeToArray(arr, data_size); }));
	GO_ERROR_IF_FAILED(benchmark(msg_type + "Serialize to memory stream ", data_size,
		[&msg, &o_mem_stream, &o_str_stream]() -> int
		{ o_str_stream.seekp(std::ios::beg); return msg.SerializeToZeroCopyStream(&o_mem_stream); }
	));
	GO_ERROR_IF_FAILED(benchmark(msg_type + "Serialize to /dev/null with FileOutputStream ", data_size,
		[&msg, &null_stream]() -> int { return msg.SerializeToZeroCopyStream(&null_stream); }));
	GO_ERROR_IF_FAILED(benchmark(msg_type + "Serialize to /dev/null reusing FileOutputStream ", data_size,
		[&msg, &reuse_null_stream]() -> int { return msg.SerializeToCodedStream(&reuse_null_stream); }));

	GO_ERROR_IF_FAILED(benchmark(msg_type + "Deserialize from byte string ", data_size,
		[&msg, &str]() -> int { return msg.ParseFromString(str); }));
	GO_ERROR_IF_FAILED(benchmark(msg_type + "Deserialize from byte array ", data_size,
		[&msg, arr, data_size]() -> int { return msg.ParseFromArray(arr, data_size); }));
	GO_ERROR_IF_FAILED(benchmark(msg_type + "Deserialize from memory stream ", data_size,
		[&msg, &i_mem_stream, &i_str_stream, &str]() -> int {
			i_str_stream.str(str);
			i_str_stream.clear();
			return msg.ParseFromZeroCopyStream(&i_mem_stream);
		}));

	close(fd);
	delete []arr;
	return 0;

error:
	cerr << "error" << endl;
	close(fd);
	delete []arr;
	return -1;
}

int main(int argc, char* argv[]) {
    GOOGLE_PROTOBUF_VERIFY_VERSION;
    if (argc != 3) {
        cerr << "Usage: " << argv[0] << " <message_dat_file1> <message_dat_file2>" << endl;
        return -1;
    }

    int ret = 0;

    ret |= run_test<benchmarks::SpeedMessage1>(argv[1], "speed1 ");
    ret |= run_test<benchmarks::SizeMessage1>(argv[1], "size1 ");

    ret |= run_test<benchmarks::SpeedMessage2>(argv[2], "speed2 ");
    ret |= run_test<benchmarks::SizeMessage2>(argv[2], "size2 ");

    return ret;
}

File: benchmarks/ProtoBench.java
// Protocol Buffers - Google's data interchange format
// Copyright 2009 Google Inc.  All rights reserved.
// https://developers.google.com/protocol-buffers/
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package com.google.protocolbuffers;

import com.google.protobuf.ByteString;
import com.google.protobuf.CodedInputStream;
import com.google.protobuf.CodedOutputStream;
import com.google.protobuf.ExtensionRegistry;
import com.google.protobuf.Message;
import com.google.protobuf.benchmarks.Benchmarks.BenchmarkDataset;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.List;

public class ProtoBench {

  private static final long MIN_SAMPLE_TIME_MS = 1 * 1000;
  private static final long TARGET_TIME_MS = 5 * 1000;

  private ProtoBench() {
    // Prevent instantiation
  }
  
  public static void main(String[] args) {
    if (args.length < 1) {
      System.err.println("Usage: ./java-benchmark <input data>");
      System.err.println("input data is in the format of \"benchmarks.proto\"");
      System.exit(1);
    }
    boolean success = true;
    for (int i = 0; i < args.length; i++) {
      success &= runTest(args[i]);
    }
    System.exit(success ? 0 : 1);
  }
  
  public static ExtensionRegistry getExtensionsRegistry(BenchmarkDataset benchmarkDataset) {
    ExtensionRegistry extensions = ExtensionRegistry.newInstance();
    if (benchmarkDataset.getMessageName().equals("benchmarks.google_message3.GoogleMessage3")) {
      benchmarks.google_message3.BenchmarkMessage38.registerAllExtensions(extensions);
      benchmarks.google_message3.BenchmarkMessage37.registerAllExtensions(extensions);
      benchmarks.google_message3.BenchmarkMessage36.registerAllExtensions(extensions);
      benchmarks.google_message3.BenchmarkMessage35.registerAllExtensions(extensions);
      benchmarks.google_message3.BenchmarkMessage34.registerAllExtensions(extensions);
      benchmarks.google_message3.BenchmarkMessage33.registerAllExtensions(extensions);
      benchmarks.google_message3.BenchmarkMessage32.registerAllExtensions(extensions);
      benchmarks.google_message3.BenchmarkMessage31.registerAllExtensions(extensions);
      benchmarks.google_message3.BenchmarkMessage3.registerAllExtensions(extensions);
    } else if (benchmarkDataset.getMessageName().equals(
        "benchmarks.google_message4.GoogleMessage4")) {
      benchmarks.google_message4.BenchmarkMessage43.registerAllExtensions(extensions);
      benchmarks.google_message4.BenchmarkMessage42.registerAllExtensions(extensions);
      benchmarks.google_message4.BenchmarkMessage41.registerAllExtensions(extensions);
      benchmarks.google_message4.BenchmarkMessage4.registerAllExtensions(extensions);      
    }
    
    return extensions;
  }

  /**
   * Return an message instance for one specific dataset, and register the extensions for that 
   * message. 
   */
  public static Message registerBenchmarks(BenchmarkDataset benchmarkDataset) {
    if (benchmarkDataset.getMessageName().equals("benchmarks.proto3.GoogleMessage1")) {
      return com.google.protobuf.benchmarks.BenchmarkMessage1Proto3.GoogleMessage1
          .getDefaultInstance();
    } else if (benchmarkDataset.getMessageName().equals("benchmarks.proto2.GoogleMessage1")) {
      return com.google.protobuf.benchmarks.BenchmarkMessage1Proto2.GoogleMessage1
          .getDefaultInstance();
    } else if (benchmarkDataset.getMessageName().equals("benchmarks.proto2.GoogleMessage2")) {
      return com.google.protobuf.benchmarks.BenchmarkMessage2.GoogleMessage2.getDefaultInstance();
    } else if (benchmarkDataset.getMessageName().
        equals("benchmarks.google_message3.GoogleMessage3")) {
      return benchmarks.google_message3.BenchmarkMessage3.GoogleMessage3.getDefaultInstance();
    } else if (benchmarkDataset.getMessageName().
        equals("benchmarks.google_message4.GoogleMessage4")) {
      return benchmarks.google_message4.BenchmarkMessage4.GoogleMessage4.getDefaultInstance();
    } else {
      return null;
    }
  }

  /**
   * Runs a single test. Error messages are displayed to stderr, and the return value indicates
   * general success/failure.
   */
  public static boolean runTest(String file) {
    final Message defaultMessage;
    BenchmarkDataset benchmarkDataset;
    ExtensionRegistry extensions;
    final byte[] inputData;

    try {
      inputData = readAllBytes(file);
      benchmarkDataset = BenchmarkDataset.parseFrom(inputData);
    } catch (IOException e) {
      System.err.println("Unable to get input data");
      return false;
    }
    defaultMessage = registerBenchmarks(benchmarkDataset);
    extensions = getExtensionsRegistry(benchmarkDataset);
    if (defaultMessage == null) {
      System.err.println("Unable to get default message " + benchmarkDataset.getMessageName());
      return false;
    }
    System.out.println("Benchmarking " + benchmarkDataset.getMessageName() + " with file " + file);

    try {
      List<byte[]> inputDataList = new ArrayList<byte[]>();
      List<ByteArrayInputStream> inputStreamList = new ArrayList<ByteArrayInputStream>();
      List<ByteString> inputStringList = new ArrayList<ByteString>();
      List<Message> sampleMessageList = new ArrayList<Message>();

      for (int i = 0; i < benchmarkDataset.getPayloadCount(); i++) {
        byte[] singleInputData = benchmarkDataset.getPayload(i).toByteArray();
        inputDataList.add(benchmarkDataset.getPayload(i).toByteArray());
        inputStreamList.add(new ByteArrayInputStream(benchmarkDataset.getPayload(i).toByteArray()));
        inputStringList.add(benchmarkDataset.getPayload(i));
        sampleMessageList.add(
            defaultMessage.newBuilderForType().mergeFrom(singleInputData, extensions).build());
      }

      FileOutputStream devNullTemp = null;
      CodedOutputStream reuseDevNullTemp = null;
      try {
        devNullTemp = new FileOutputStream("/dev/null");
        reuseDevNullTemp = CodedOutputStream.newInstance(devNullTemp);
      } catch (FileNotFoundException e) {
        // ignore: this is probably Windows, where /dev/null does not exist
      }
      final FileOutputStream devNull = devNullTemp;
      final CodedOutputStream reuseDevNull = reuseDevNullTemp;
      benchmark(
          "Serialize to byte string",
          inputData.length,
          new Action() {
            public void execute() {
              for (int i = 0; i < sampleMessageList.size(); i++) {
                sampleMessageList.get(i).toByteString();
              }
            }
          });
      benchmark(
          "Serialize to byte array",
          inputData.length,
          new Action() {
            public void execute() {
              for (int i = 0; i < sampleMessageList.size(); i++) {
                sampleMessageList.get(i).toByteString();
              }
            }
          });
      benchmark(
          "Serialize to memory stream",
          inputData.length,
          new Action() {
            public void execute() throws IOException {
              ByteArrayOutputStream output = new ByteArrayOutputStream();
              for (int i = 0; i < sampleMessageList.size(); i++) {
                sampleMessageList.get(i).writeTo(output);
              }
            }
          });
      if (devNull != null) {
        benchmark(
            "Serialize to /dev/null with FileOutputStream",
            inputData.length,
            new Action() {
              public void execute() throws IOException {
                for (int i = 0; i < sampleMessageList.size(); i++) {
                  sampleMessageList.get(i).writeTo(devNull);
                }
              }
            });
        benchmark(
            "Serialize to /dev/null reusing FileOutputStream",
            inputData.length,
            new Action() {
              public void execute() throws IOException {
                for (int i = 0; i < sampleMessageList.size(); i++) {
                  sampleMessageList.get(i).writeTo(reuseDevNull);
                  reuseDevNull.flush(); // force the write to the OutputStream
                }
              }
            });
      }
      benchmark(
          "Deserialize from byte string",
          inputData.length,
          new Action() {
            public void execute() throws IOException {
              for (int i = 0; i < inputStringList.size(); i++) {
                defaultMessage
                    .newBuilderForType()
                    .mergeFrom(inputStringList.get(i), extensions)
                    .build();
              }
            }
          });
      benchmark(
          "Deserialize from byte array",
          inputData.length,
          new Action() {
            public void execute() throws IOException {
              for (int i = 0; i < inputDataList.size(); i++) {
                defaultMessage
                    .newBuilderForType()
                    .mergeFrom(CodedInputStream.newInstance(inputDataList.get(i)), extensions)
                    .build();
              }
            }
          });
      benchmark(
          "Deserialize from memory stream",
          inputData.length,
          new Action() {
            public void execute() throws IOException {
              for (int i = 0; i < inputStreamList.size(); i++) {
                defaultMessage
                    .newBuilderForType()
                    .mergeFrom(CodedInputStream.newInstance(inputStreamList.get(i)), extensions)
                    .build();
                inputStreamList.get(i).reset();
              }
            }
          });
      System.out.println();
      return true;
    } catch (Exception e) {
      System.err.println("Error: " + e.getMessage());
      System.err.println("Detailed exception information:");
      e.printStackTrace(System.err);
      return false;
    }
  }
  
  private static void benchmark(String name, long dataSize, Action action) throws IOException {
    // Make sure it's JITted "reasonably" hard before running the first progress test
    for (int i=0; i < 100; i++) {
      action.execute();
    }
    
    // Run it progressively more times until we've got a reasonable sample
    int iterations = 1;
    long elapsed = timeAction(action, iterations);
    while (elapsed < MIN_SAMPLE_TIME_MS) {
      iterations *= 2;
      elapsed = timeAction(action, iterations);
    }
    
    // Upscale the sample to the target time. Do this in floating point arithmetic
    // to avoid overflow issues.
    iterations = (int) ((TARGET_TIME_MS / (double) elapsed) * iterations);
    elapsed = timeAction(action, iterations);
    System.out.println(name + ": " + iterations + " iterations in "
         + (elapsed/1000f) + "s; " 
         + (iterations * dataSize) / (elapsed * 1024 * 1024 / 1000f) 
         + "MB/s");
  }
  
  private static long timeAction(Action action, int iterations) throws IOException {
    System.gc();    
    long start = System.currentTimeMillis();
    for (int i = 0; i < iterations; i++) {
      action.execute();
    }
    long end = System.currentTimeMillis();
    return end - start;
  }
  
  private static byte[] readAllBytes(String filename) throws IOException {
    RandomAccessFile file = new RandomAccessFile(new File(filename), "r");
    byte[] content = new byte[(int) file.length()];
    file.readFully(content);
    return content;
  }

  /**
   * Interface used to capture a single action to benchmark.
   */
  interface Action {
    void execute() throws IOException;
  }
}
